contracts/universalResolver/UniversalResolver.sol

```diff
@@ -0,0 +1,349 @@
+            address /*resolver*/,
+            address /*reverseResolver*/
+        )
+    {
+        // https://docs.ens.domains/ensip/19
+        Lookup memory lookup = lookupResolver(
+            NameCoder.encode(ENSIP19.reverseName(encodedAddress, coinType))
+        );
+        bytes memory v = _resolveBatch(
+            lookup,
+            _oneCall(abi.encodeCall(INameResolver.name, (lookup.node))),
+            gateways,
+            this.reverseNameCallback.selector,
+            abi.encode(ReverseArgs(encodedAddress, coinType, gateways))
+        );
+        assembly {
+            return(add(v, 32), mload(v))
+        }
+    }
+
+    struct ReverseArgs {
+        bytes encodedAddress;
+        uint256 coinType;
+        string[] gateways;
+    }
+
+    function reverseNameCallback(
+        Lookup calldata revLookup,
+        Thread[] calldata threads,
+        bytes memory v // stack too deep
+    )
+        external
+        view
+        returns (
+            string memory primary,
+            address /*resolver*/,
+            address /*reverseResolver*/
+        )
+    {
+        ReverseArgs memory args = abi.decode(v, (ReverseArgs));
+        v = _requireResponse(threads[0]);
+        primary = abi.decode(v, (string));
+        if (bytes(primary).length == 0) {
+            return ("", address(0), revLookup.resolver); // name() was empty
+        }
+        Lookup memory lookup = lookupResolver(NameCoder.encode(primary));
+        v = _resolveBatch(
+            lookup,
+            _oneCall(
+                args.coinType == COIN_TYPE_ETH
+                    ? abi.encodeCall(IAddrResolver.addr, (lookup.node))
+                    : abi.encodeCall(
+                        IAddressResolver.addr,
+                        (lookup.node, args.coinType)
+                    )
+            ),
+            args.gateways,
+            this.reverseAddressCallback.selector,
+            abi.encode(args.encodedAddress, primary, revLookup.resolver)
+        );
+        assembly {
+            return(add(v, 32), mload(v))
+        }
+    }
+
+    function reverseAddressCallback(
+        Lookup calldata lookup,
+        Thread[] calldata threads,
+        bytes calldata myData
+    )
+        external
+        pure
+        returns (
+            string memory primary,
+            address resolver,
+            address reverseResolver
+        )
+    {
+        bytes memory reverseAddress;
+        (reverseAddress, primary, reverseResolver) = abi.decode(
+            myData,
+            (bytes, string, address)
+        );
+        bytes memory v = _requireResponse(threads[0]);
+        bytes memory primaryAddress;
+        if (bytes4(threads[0].call) == IAddrResolver.addr.selector) {
+            address addr = abi.decode(v, (address));
+            if (addr != address(0)) {
+                primaryAddress = abi.encodePacked(addr);
+            }
+        } else {
+            primaryAddress = abi.decode(v, (bytes));
+        }
```

Comment:
I assume this is assuming it must be `IAddressResolver.addr.selector`, but we should explicitly check.